src/objectExplorer/objectExplorerService.ts

```diff
@@ -56,70 +107,155 @@ export class ObjectExplorerService {
         const self = this;
         const handler = (result: ExpandResponse) => {
             if (result && result.nodes) {
-                const children = result.nodes.map(node => TreeNodeInfo.fromNodeInfo(node, self.currentNode.sessionId, self.currentNode));
-                self._currentNode.collapsibleState = TreeItemCollapsibleState.Expanded;
-                self._treeNodeToChildrenMap.set(self.currentNode, children);
-                return self._objectExplorerProvider.refresh(self.currentNode);
+                const children = result.nodes.map(node => TreeNodeInfo.fromNodeInfo(node, self._currentNode.sessionId,
+                    self._currentNode, self._currentNode.connectionCredentials));
+                self._treeNodeToChildrenMap.set(self._currentNode, children);
+                const expandParams: ExpandParams = {
+                    sessionId: result.sessionId,
+                    nodePath: result.nodePath
+                }
+                for (let key of self._expandParamsToPromiseMap.keys()) {
+                    if (key.sessionId === expandParams.sessionId &&
+                        key.nodePath === expandParams.nodePath) {
+                        let promise = self._expandParamsToPromiseMap.get(key);
+                        return promise.resolve(children);
+                    }
+                }
             }
         };
         return handler;
     }
 
-    private async expandNode(node: TreeNodeInfo, sessionId: string): Promise<boolean> {
+    private async expandNode(node: TreeNodeInfo, sessionId: string, promise: Deferred<TreeNodeInfo[]>): Promise<boolean> {
         const expandParams: ExpandParams = {
             sessionId: sessionId,
             nodePath: node.nodePath
         };
         const response = await this._connectionManager.client.sendRequest(ExpandRequest.type, expandParams);
+        if (promise) {
+            this._expandParamsToPromiseMap.set(expandParams, promise);
+        }
         return response;
     }
 
-    async getChildren(element?: TreeNodeInfo): Promise<TreeNodeInfo[]> {
+    private updateNode(node: TreeNodeInfo): void {
+        for (let rootTreeNode of this._rootTreeNodeArray) {
+            if (rootTreeNode.connectionCredentials === node.connectionCredentials &&
+                rootTreeNode.label === node.label) {
+                    const index = this._rootTreeNodeArray.indexOf(rootTreeNode);
+                    this._rootTreeNodeArray[index] = node;
+                    return;
+            }
+        }
+        this._rootTreeNodeArray.push(node);
+    }
+
+    async getChildren(element?: TreeNodeInfo): Promise<vscode.TreeItem[]> {
```

Comment:
>async [](start = 4, length = 5)

All functions should have visibility modifiers so it's explicit what it should be